src/app/chat/chat.component.ts

```diff
@@ -27,34 +27,60 @@ export class ChatComponent {
     userId: string | null;
     username: string | null;
   }> = this.store.select(selectUserData);
-  protected readonly datasource: IDatasource<Message> = this.getDatasource();
+
+  protected readonly datasource = new Datasource<Message>({
+    get: (index: number, count: number) => {
+      console.log(index, count);
+      return this.store
+        .select(selectMessagesSlice(index, count))
+        .pipe(take(1), tap(console.log));
+    },
+    devSettings: { debug: true },
+  });
 
   constructor(
     private fb: FormBuilder,
     private store: Store<EntityState<Message>>
   ) {
+    let prev: Message[] = [];
     this.messages$
       .pipe(
-        tap(messages =>
-          this.datasource.adapter?.reset(this.getDatasource(messages.length))
-        )
+        tap(async curr => {
+          const newItem = curr.find(c => !prev.find(p => c.id == p.id));
+          if (newItem) {
+            prev = curr;
+            this.appendMessage(newItem);
+            return;
+          }
+          const oldItem = curr.find(c =>
+            prev.find(p => c.id == p.id && c.time !== p.time)
+          );
+          if (oldItem) {
+            prev = curr;
+            this.replaceMessage(oldItem);
+            return;
+          }
+        })
       )
       .subscribe();
   }
 
-  getDatasource<T>(maxIndex?: number): IDatasource<T> {
-    return new Datasource<T>({
-      get: (index: number, count: number) => {
-        // why is index always 1?
-        // why does index start at 1?
-        console.log(index, count);
-        return this.store
-          .select(selectMessagesSlice(index, count))
-          .pipe(take(1), tap(console.log));
-      },
-      settings: {
-        maxIndex: maxIndex ?? Infinity,
-      },
+  async appendMessage(message: Message) {
+    await this.datasource.adapter.relax();
+    await this.datasource.adapter.append({
+      items: [message],
+      eof: true,
+    });
+    await this.datasource.adapter.clip({
+      forwardOnly: true,
+    });
+  }
+
+  async replaceMessage(message: Message) {
+    await this.datasource.adapter.relax();
+    await this.datasource.adapter.replace({
+      predicate: ({ data }) => data.id === message.id,
+      items: [message],
```

Comment:
This provides hard replace: item is being removed and re-created again with new contents. In case of replacing virtual items this doesn't affect the UI. But it replaces in-buffer items, I would say, there may be a place for optimization, and perhaps you may want to look at the [adapter-fix-updater](https://dhilt.github.io/ngx-ui-scroll/#/experimental#adapter-fix-updater) API...

Reply:
I just tested this locally, and once items are added to the DOM (I tried over 30 items) they are never removed. Is this expected? 🤔 I thought only items in the default buffer size (5) would be rendered

Reply:
if adapter-fix-updater is needed, is there a way to run the updater function only for the item I want? I'm thinking that iterating over the whole collection to update one item might result in poor performance if you have thousands of messages in the collection

Reply:
Predicate's traversing only the in-buffer (non-virtual) items that are present in the DOM, not whole collection. It can't hit performance.

On items remove, I pushed "clip" fix, now it should properly clip in all cases. The issue with the previous revision was that I tested the UI by scrolling to the very bottom each time before a new adding. So if you didn't scroll before new append, it remained unclipped. It is fixed now.